--- a/controller/pagecontroller.php
+++ b/controller/pagecontroller.php
@@ -26,6 +26,8 @@
 	public function __construct($appName, IRequest $request, $userId) {
 		parent::__construct($appName, $request);
 
+		\OC::$loader->addValidRoot("/etc/spreedme/owncloud");
+
 		Helper::notifyIfAppNotSetUp();
 
 		$this->userid = $userId;
--- a/appinfo/app.php
+++ b/appinfo/app.php
@@ -14,6 +14,8 @@
 use OCA\SpreedME\Settings\Settings;
 use OCP\AppFramework\App;
 
+\OC::$loader->addValidRoot("/etc/spreedme/owncloud");
+
 $app = new App(Settings::APP_ID);
 $container = $app->getContainer();
 
--- a/controller/apicontroller.php
+++ b/controller/apicontroller.php
@@ -27,6 +27,8 @@
 	public function __construct($appName, IRequest $request, $userId, IURLGenerator $urlGenerator) {
 		parent::__construct($appName, $request);
 
+		\OC::$loader->addValidRoot("/etc/spreedme/owncloud");
+
 		if (!empty($userId)) {
 			$this->user = new User($userId);
 		} else {
